sftp/sftp.go

```diff
@@ -0,0 +1,228 @@
+	missingLoggerInterface           = "Logger interface is not defined"
+	invalidPort                      = "Port must not be 0 or bigger then 65535"
+
+	// ServerStarted is logged when the SFTP server is first launched.
+	ServerStarted LogType = "server-started"
+	// ChannelNotSession is logged when the SFTP receives a request for a new channel which is NOT of type 'session'.
+	ChannelNotSession LogType = "channel-not-session"
+	// AcceptNetworkError is logged when there is an error accepting network connections within the listener.
+	AcceptNetworkError LogType = "accept-network-error"
+	// SSHKeyExchangeError is logged when there is an error performing a key exchange between the SFTP client and server.
+	SSHKeyExchangeError LogType = "ssh-key-exchange-error"
+	// AcceptChannelError is logged when there is an error while trying to accept the new request channel.
+	AcceptChannelError LogType = "accept-channel-error"
+)
+
+type logger interface {
+	Info(tag LogType, msg string)
+	Error(tag LogType, err error)
+}
+
+// Server implements a composable SFTP Server.
+type Server struct {
+	port                 int
+	publicIP             string
+	sshConfig            ssh.ServerConfig
+	sshHandshakeDeadline time.Duration
+	logger               logger
+	beforeHandle         func(conn net.Conn, err error) bool
+	handleSFTPSession    func(channel ssh.Channel, sconn *ssh.ServerConn)
+	listener             net.Listener
+}
+
+// Options defines required configurations
+// used when calling NewServer().
+type Options struct {
+	Port      int
+	PublicIP  string
+	Logger    logger
+	SSHConfig *ssh.ServerConfig
+	// SSHHandshakeDeadline controls the time.Duration which ssh session
+	// have to complete their handshake. This option is not a part of the
+	// ssh.ServerConfig so we had to implement it separately.
+	SSHHandshakeDeadline time.Duration
+	// BeforeHandle will be executed before `HandleSFTPSession` and before
+	// error checking happens during the socket listener.Accept().
+	BeforeHandle func(conn net.Conn, err error) bool
+	// HandleSFTPSession is executed when a new SFTP session is requested.
+	HandleSFTPSession func(channel ssh.Channel, sconn *ssh.ServerConn)
+}
+
+func defaultBeforeAccept(_ net.Conn, _ error) (accept bool) {
+	return true
+}
+
+// NewServer composes a new Server{} object from the options given.
+//
+// It is recommended to use (2*time.Minute) as the SSHHandshakeDeadline.
+// 2 minutes is the default deadline for OpenSSH servers/clients.
+func NewServer(options *Options) (sftpServer *Server, err error) {
+	if options.HandleSFTPSession == nil {
+		return nil, errors.New(missingConnectionHandlerFunction)
+	}
+	if options.SSHConfig == nil {
+		return nil, errors.New(missingSSHConfig)
+	}
+	if options.Logger == nil {
+		return nil, errors.New(missingLoggerInterface)
+	}
+	if options.Port < 1 || options.Port > 65535 {
+		return nil, errors.New(invalidPort)
+	}
+	if options.BeforeHandle == nil {
+		options.BeforeHandle = defaultBeforeAccept
```

Comment:
Can't it just remain nil and you check before you use it?

Reply:
It's used every time we accept a connection, so I figured a single check would be better then multiple.

Reply:
Without checking you are setting up the function call and calling+returning every time - since it is a function pointer it can never be skipped however if we do:

`if options.BeforeHandle != nil && !options.BeforeHandle(....) {`

... the only cost will be the nil check - which the compiler does anyway if you don't add it.

Since it is a fully predictable branch you really aren't leaving anything on the table.

Reply:
you are right, I didn't even think about that option. I'll change it right away